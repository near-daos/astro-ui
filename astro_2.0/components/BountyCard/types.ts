import { Token } from 'types/token';
import { BountyStatus } from 'types/bounties';

export enum CardType {
  Bounty = 'Bounty',
  Claim = 'Claim',
}
export interface BountyCardContent {
  id: string;
  daoId: string;
  amount: string;
  type: CardType;
  description: string;
  externalUrl?: string;
  token: Token;
  status: BountyStatus;
  bountyBond: string;
  slots?: number;
  slotsTotal?: number;
  claimedByCurrentUser: boolean;
  accountId?: string;
  timeToComplete: string;
  forgivenessPeriod: string;
  relatedProposalHash?: string;
}
