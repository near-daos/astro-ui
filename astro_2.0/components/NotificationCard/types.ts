import { NotificationType, NotificationStatus } from 'types/notification';

export interface NotificationCardContent {
  id: string;
  type: NotificationType;
  status: NotificationStatus;
  text: string;
  time: string;
  flagCover?: string;
  logo?: string;
  url?: string;
}
