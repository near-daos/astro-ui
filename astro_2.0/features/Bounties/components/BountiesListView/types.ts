import { Bounty } from 'types/bounties';
import { ReactNode } from 'react';
import { ProposalVariant } from 'types/proposal';

export type SectionItem = {
  id: string;
  title: string;
  proposer: string;
  proposalId: string;
  bounty?: Bounty;
  link: {
    pathname: string;
    query: {
      dao: string;
      bounty?: string;
      proposal?: string;
    };
  };
  completeHandler?: (
    id: number,
    variant: ProposalVariant.ProposeDoneBounty
  ) => void;
  content: ReactNode;
};
