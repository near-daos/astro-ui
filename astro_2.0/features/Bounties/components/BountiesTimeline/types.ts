export type TimelineMilestoneType =
  | 'Proposal Created'
  | 'Bounty Created'
  | 'Claim'
  | 'Pending Approval'
  | 'Complete Claim'
  | 'Claim Deadline'
  | 'Complete Bounty';

export type TimelineMilestone = {
  type: TimelineMilestoneType;
  date: Date;
  tooltip: string;
  color?: string;
};

export type TimelineClaim = {
  id: string;
  milestones: TimelineMilestone[];
  title: string;
  color: string;
  minDate?: Date;
  maxDate?: Date;
};

export type TimelineGroup = {
  id: string;
  name: string;
  isOpen: boolean;
  milestones: TimelineMilestone[];
  claims: TimelineClaim[];
  minDate?: Date;
  maxDate?: Date;
};

export type TimelineGranularity = 'month' | 'day' | 'hour';
