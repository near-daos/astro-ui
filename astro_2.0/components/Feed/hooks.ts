import { useAsyncFn, useList, useLocalStorage, useLocation } from 'react-use';
import { useCallback } from 'react';
import { VoteAction } from 'types/proposal';
import { useWalletContext } from 'context/WalletContext';
import { useRouter } from 'next/router';
import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useMultiVoteActions(): {
  loading: boolean;
  handleVote: (vote: VoteAction) => Promise<void>;
  handleSelect: (id: string) => void;
  handleDismiss: () => void;
  list: string[];
} {
  const { nearService } = useWalletContext();
  const router = useRouter();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [list, { push, removeAt, clear }] = useList<string>([]);

  const handleSelect = useCallback(
    id => {
      const itemIndex = list.findIndex(item => item === id);

      if (itemIndex !== -1) {
        removeAt(itemIndex);
      } else {
        push(id);
      }
    },
    [list, push, removeAt]
  );

  const handleDismiss = useCallback(() => {
    clear();
  }, [clear]);

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      try {
        const paramsArr = list.map(item => {
          const separatorIndex = item.lastIndexOf('-');
          const daoId = item.substring(0, separatorIndex);
          const proposalId = item.substring(separatorIndex + 1);

          return { daoId, proposalId };
        });

        setVoteActionSource(pathname);

        await nearService?.multiVote(vote, paramsArr);

        await router.reload();
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          description: e.message,
          lifetime: 20000,
        });

        await router.reload();
      }
    },
    [router, nearService, list]
  );

  return {
    loading,
    list,
    handleVote,
    handleSelect,
    handleDismiss,
  };
}
