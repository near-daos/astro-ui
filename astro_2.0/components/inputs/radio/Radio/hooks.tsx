/* eslint-disable import/prefer-default-export */
import { useContext } from 'react';

import {
  RadioContext,
  RadioContextType,
} from 'astro_2.0/components/inputs/radio/RadioGroup';

export const useRadioContext = (): RadioContextType => {
  const context = useContext(RadioContext);

  if (!context) {
    throw new Error(
      'Radio compound components cannot be rendered outside the Radio component'
    );
  }

  return context;
};
