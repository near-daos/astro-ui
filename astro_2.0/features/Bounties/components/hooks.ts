import { useCallback, useRef, useState } from 'react';
import { useRouter } from 'next/router';
import { useAsyncFn, useLocalStorage, useLocation } from 'react-use';
import axios, { CancelTokenSource } from 'axios';

import { SputnikHttpService } from 'services/sputnik';
import { useModal } from 'components/modal';
import { ConfirmActionModal } from 'astro_2.0/components/ConfirmActionModal';

import { Bounty, BountyContext, BountyProposal } from 'types/bounties';
import { VoteAction } from 'types/proposal';

import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';

import { useWalletContext } from 'context/WalletContext';
import { PaginationResponse } from 'types/api';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useBountyControls(
  daoId: string,
  bounty?: Bounty
): {
  handleClaim: () => void;
  handleUnclaim: () => void;
  loading: boolean;
} {
  const router = useRouter();
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [loading, setLoading] = useState(false);

  const [showModal] = useModal(ConfirmActionModal);

  const onSuccessHandler = useCallback(async () => {
    router.reload();
  }, [router]);

  const handleClaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Claim',
    });

    if (res?.length && bounty) {
      setLoading(true);

      const dao = await SputnikHttpService.getDaoById(daoId);

      if (!dao) {
        return;
      }

      try {
        setVoteActionSource(pathname);

        await nearService?.claimBounty(daoId, {
          bountyId: bounty?.bountyId,
          deadline: bounty?.maxDeadline,
          bountyBond: dao.policy.bountyBond,
          gas: res[0],
          tokenId: bounty?.token,
        });

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    daoId,
    setVoteActionSource,
    pathname,
    nearService,
    onSuccessHandler,
  ]);

  const handleUnclaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Unclaim',
    });

    if (res?.length && bounty) {
      try {
        setLoading(true);
        setVoteActionSource(pathname);
        await nearService?.unclaimBounty(daoId, bounty?.bountyId);

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    setVoteActionSource,
    pathname,
    nearService,
    daoId,
    onSuccessHandler,
  ]);

  return {
    handleClaim,
    handleUnclaim,
    loading,
  };
}

export function useBountyVoting(
  daoId: string,
  proposal: BountyProposal
): {
  handleVote: (vote: VoteAction) => void;
  loading: boolean;
} {
  const router = useRouter();
  const [showModal] = useModal(ConfirmActionModal);
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      const res = await showModal({
        title: 'Confirm Your Vote',
      });

      if (res?.length) {
        setVoteActionSource(pathname);
        await nearService?.vote(daoId, proposal.proposalId, vote, res[0]);
        await router.reload();
      }
    },
    [daoId, proposal, router, nearService, pathname, setVoteActionSource]
  );

  return {
    handleVote,
    loading,
  };
}

export function useBountySearch(): {
  handleSearch: (
    val: string
  ) => Promise<PaginationResponse<BountyContext[]> | null>;
  loading: boolean;
} {
  const router = useRouter();
  const daoId = router.query.dao as string;
  const { accountId } = useWalletContext();
  const cancelTokenRef = useRef<CancelTokenSource | null>(null);

  const [{ loading }, handleSearch] = useAsyncFn(async query => {
    if (cancelTokenRef.current) {
      cancelTokenRef.current?.cancel('Cancelled by new req');
    }

    const { CancelToken } = axios;
    const source = CancelToken.source();

    cancelTokenRef.current = source;

    return SputnikHttpService.findBountyContext({
      daoId: daoId || '',
      accountId,
      query,
      cancelToken: source.token,
    });
  }, []);

  return {
    handleSearch,
    loading,
  };
}

export function useHideBounty(): {
  handleSubmit: () => void;
  handleSelect: (id: string) => void;
  selected: string[];
  loading: boolean;
} {
  const router = useRouter();
  const showHidden = router.query?.bountyFilter === 'hidden';
  const daoId = router.query.dao as string;
  const { accountId, nearService, pkAndSignature } = useWalletContext();
  const [selected, setSelected] = useState<string[]>([]);

  const [{ loading }, handleSubmit] = useAsyncFn(async () => {
    if (!pkAndSignature) {
      return;
    }

    try {
      const { publicKey, signature } = pkAndSignature;

      if (publicKey && signature) {
        await SputnikHttpService.toggleBountyContexts({
          accountId,
          publicKey,
          signature,
          daoId,
          ids: selected,
          isArchived: !showHidden,
        });

        await router.replace(router.asPath, undefined, {
          shallow: false,
        });

        setSelected([]);
      }
    } catch (err) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: err.message,
      });
    }
  }, [
    selected,
    daoId,
    showHidden,
    accountId,
    router,
    nearService,
    pkAndSignature,
  ]);

  const handleSelect = useCallback(
    (id: string) => {
      if (selected.includes(id)) {
        setSelected(selected.filter(item => item !== id));
      } else {
        setSelected([...selected, id]);
      }
    },
    [selected]
  );

  return {
    handleSubmit,
    handleSelect,
    selected,
    loading,
  };
}
