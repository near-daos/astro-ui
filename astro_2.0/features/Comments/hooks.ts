import { useCallback, useEffect, useState } from 'react';
import { useRouter } from 'next/router';
import uniqBy from 'lodash/uniqBy';
import { useAsyncFn, useMount, useMountedState } from 'react-use';
import { useDraftsContext } from 'astro_2.0/features/Drafts/components/DraftsProvider';
import { DraftComment } from 'services/DraftsService/types';
import { useWalletContext } from 'context/WalletContext';
import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';
import { configService } from 'services/ConfigService';
import io, { Socket as TSocket } from 'socket.io-client';
import { fetcher as getDraftProposalComments } from 'services/ApiService/hooks/useDraftComments';
import { useFlags } from 'launchdarkly-react-client-sdk';
import { DraftCommentIndex } from 'services/SearchService/types';
import { mapDraftCommentIndexToDraftComment } from 'services/SearchService/mappers/draftComment';

type Socket = typeof TSocket;

function prepareData(data: DraftComment[]) {
  const groups = data.reduce<Record<string, DraftComment[]>>((res, item) => {
    const key = item.replyTo ?? '';

    if (res[key]) {
      res[key].push(item);
    } else {
      res[key] = [item];
    }

    return res;
  }, {});

  if (!groups['']) {
    return data;
  }

  return groups[''].map(item => {
    return {
      ...item,
      replies: groups[item.id] ?? [],
    };
  });
}

export function useDraftComments(): {
  loading: boolean;
  countComments: number;
  data: DraftComment[];
  addComment: (val: string) => Promise<void>;
  editComment: (val: string, id: string) => Promise<void>;
  deleteComment: (id: string) => Promise<void>;
  likeComment: (id: string, isLiked: boolean) => Promise<void>;
  dislikeComment: (id: string, idDislike: boolean) => Promise<void>;
} {
  const isMounted = useMountedState();
  const router = useRouter();
  const { draft, dao } = router.query;
  const contextId = draft as string;
  const daoId = dao as string;
  const contextType = 'DraftProposal';

  const { useOpenSearchDataApi } = useFlags();
  const { draftsService, setAmountComments } = useDraftsContext();
  const { accountId, pkAndSignature } = useWalletContext();

  const [value, setValue] = useState<{
    data: DraftComment[];
    countComments: number;
  }>({
    data: [],
    countComments: 0,
  });

  const [{ loading }, getAllComments] = useAsyncFn(async () => {
    if (useOpenSearchDataApi === undefined) {
      return;
    }

    try {
      const data = useOpenSearchDataApi
        ? await getDraftProposalComments(
            'draftProposalComment',
            daoId,
            contextId
          )
        : await draftsService.getDraftComments({
            contextId,
            contextType,
            offset: 0,
            limit: 1000,
          });

      setValue({
        data,
        countComments: data.length,
      });
    } catch (e) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: e?.message,
      });
    }
  }, [contextId, useOpenSearchDataApi]);

  useMount(() => getAllComments());

  const addComment = useCallback(
    async (msg: string, replyTo?: string) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      try {
        await draftsService.createDraftComment({
          contextId,
          daoId,
          contextType,
          message: msg,
          replyTo,
          accountId,
          publicKey,
          signature,
        });
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  const editComment = useCallback(
    async (msg: string, id: string) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      try {
        await draftsService.editDraftComment({
          id,
          daoId,
          draftId: contextId,
          message: msg,
          accountId,
          publicKey,
          signature,
        });

        await getAllComments();
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [pkAndSignature, draftsService, daoId, contextId, accountId, getAllComments]
  );

  const likeComment = useCallback(
    async (id: string, isLiked: boolean) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      const params = {
        id,
        daoId,
        draftId: contextId,
        accountId,
        publicKey,
        signature,
      };

      try {
        if (isLiked) {
          await draftsService.removeLikeDraftComment(params);
        } else {
          await draftsService.removeDislikeDraftComment(params);
          await draftsService.likeDraftComment(params);
        }
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  const dislikeComment = useCallback(
    async (id: string, isDislike: boolean) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      const params = {
        id,
        daoId,
        draftId: contextId,
        accountId,
        publicKey,
        signature,
      };

      try {
        if (isDislike) {
          await draftsService.removeDislikeDraftComment(params);
        } else {
          await draftsService.removeLikeDraftComment(params);
          await draftsService.dislikeDraftComment(params);
        }
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  const deleteComment = useCallback(
    async (id: string) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      try {
        await draftsService.deleteDraftComment({
          id,
          daoId,
          draftId: contextId,
          accountId,
          publicKey,
          signature,
        });
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  useEffect(() => {
    setAmountComments(value?.countComments || 0);
  }, [value?.countComments, setAmountComments]);

  useEffect(() => {
    let socket: Socket;
    const { appConfig } = configService.get();

    const { publicKey, signature } = pkAndSignature || {};

    if (accountId && publicKey && isMounted() && appConfig) {
      socket = io(appConfig.DRAFTS_API_URL, {
        query: {
          accountId,
          publicKey,
          signature,
        },
        transports: ['websocket'],
      });

      if (socket) {
        socket.on(
          'draft-comment',
          (comment: DraftComment | DraftCommentIndex) => {
            if (isMounted()) {
              setValue(prev => {
                const newValue = mapDraftCommentIndexToDraftComment(
                  comment as DraftCommentIndex
                ) as DraftComment;

                return {
                  data: uniqBy([newValue, ...prev.data], item => item.id),
                  countComments: prev.countComments + 1,
                };
              });
            }
          }
        );
        socket.on(
          'draft-comment-updated',
          (comment: DraftComment | DraftCommentIndex) => {
            if (isMounted()) {
              setValue(prev => {
                const newValue = mapDraftCommentIndexToDraftComment(
                  comment as DraftCommentIndex
                ) as DraftComment;

                const newData = prev.data.map(item => {
                  if (item.id !== newValue.id) {
                    return item;
                  }

                  return newValue;
                });

                return {
                  data: newData,
                  countComments: newData.length,
                };
              });
            }
          }
        );
        socket.on(
          'draft-comment-removed',
          (comment: DraftComment | DraftCommentIndex) => {
            if (isMounted()) {
              const newValue = mapDraftCommentIndexToDraftComment(
                comment as DraftCommentIndex
              ) as DraftComment;

              setValue(prev => {
                const newData = prev.data.filter(
                  item => item.id !== newValue.id
                );

                return {
                  data: newData,
                  countComments: newData.length,
                };
              });
            }
          }
        );
      }
    }
  }, [accountId, isMounted, pkAndSignature, useOpenSearchDataApi]);

  const preparedData = prepareData(value.data);

  return {
    loading,
    countComments: preparedData?.length || 0,
    data: preparedData ?? [],
    addComment,
    editComment,
    deleteComment,
    likeComment,
    dislikeComment,
  };
}
