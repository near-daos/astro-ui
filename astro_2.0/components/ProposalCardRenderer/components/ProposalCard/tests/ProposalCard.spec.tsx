/* eslint-disable @typescript-eslint/ban-ts-comment */

import { render } from 'jest/testUtils';
import { useRouter } from 'next/router';

import { SINGLE_PROPOSAL_PAGE_URL } from 'constants/routing';

import { VoteDetail } from 'features/types';
import { ProposalType, ProposalVariant } from 'types/proposal';

import {
  ProposalCard,
  ProposalCardProps,
} from 'astro_2.0/components/ProposalCardRenderer/components/ProposalCard';
import { fireEvent } from '@testing-library/dom';

import { useCountdown } from 'hooks/useCountdown';

jest.mock('next-i18next', () => ({
  // this mock makes sure any components using the translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str: string): string => str,
    };
  },
}));

jest.mock('next/router', () => {
  return {
    useRouter: jest.fn(() => ({
      asPath: '',
    })),
  };
});

jest.mock('react-use', () => {
  return {
    ...jest.requireActual('react-use'),
    useMedia: jest.fn(),
  };
});

jest.mock('components/VoteDetails/components/progress-bar/ProgressBar', () => ({
  ProgressBar: ({ detail }: { detail: VoteDetail }) => (
    <div>{detail.label}</div>
  ),
}));

jest.mock('components/Icon', () => {
  return {
    Icon: ({ name }: { name: string }) => <div>{name}</div>,
  };
});

jest.mock('hooks/useCountdown', () => ({
  useCountdown: jest.fn(),
}));

describe('Proposal Card', () => {
  function renderProposalCard(props?: Partial<ProposalCardProps>) {
    return render(
      <ProposalCard
        type={ProposalType.Vote}
        variant={ProposalVariant.ProposePoll}
        status="InProgress"
        proposer="proposer"
        description="description"
        link="link"
        proposalTxHash="proposalTxHash"
        daoId="daoId"
        proposalId={123}
        accountId="accountId"
        likes={0}
        dislikes={0}
        voteRemove={0}
        voteStatus="InProgress"
        isFinalized={false}
        liked={false}
        disliked={false}
        dismissed={false}
        content={<div>Content</div>}
        votePeriodEnd="2016-12-12"
        commentsCount={0}
        permissions={{
          canApprove: true,
          canReject: true,
          canDelete: true,
          isCouncil: true,
        }}
        {...props}
      />
    );
  }

  it('Should render component', () => {
    const { container } = renderProposalCard();

    expect(container).toMatchSnapshot();
  });

  it('Should render progress bar when vote details provided', () => {
    const label = 'I am vote details';
    const { getAllByText } = renderProposalCard({
      voteDetails: {
        limit: '3',
        label,
      },
    });

    expect(getAllByText(label)).toHaveLength(1);
  });

  it('Should render "Finalize" button', () => {
    const { getAllByText } = renderProposalCard({
      voteStatus: 'Expired',
    });

    expect(getAllByText('Finalize')).toHaveLength(1);
  });

  it.each`
    status        | icon
    ${'Approved'} | ${'sealApproved'}
    ${'Expired'}  | ${'sealFailed'}
    ${'Moved'}    | ${'sealFailed'}
    ${'Rejected'} | ${'sealFailed'}
    ${'Removed'}  | ${'sealFailed'}
  `('Should render $icon for $status status', ({ status, icon }) => {
    // @ts-ignore
    useCountdown.mockImplementation(() => '123');

    const { getAllByText } = renderProposalCard({
      votePeriodEnd: '3016-12-12',
      status,
    });

    expect(getAllByText(icon)).toHaveLength(1);
  });

  it('Should navigate to proposal page when clicked', () => {
    const id = '123456';
    const daoId = 'daoId123';

    const router = {
      push: jest.fn(),
    };

    // @ts-ignore
    useRouter.mockImplementation(() => router);

    const component = renderProposalCard({ id, daoId });

    fireEvent.click(component.getByTestId('proposal-card-root'));

    expect(router.push).toBeCalledWith({
      pathname: SINGLE_PROPOSAL_PAGE_URL,
      query: {
        dao: daoId,
        proposal: id,
      },
    });
  });

  it('Should render proper timestamp when status InProgress and time left', () => {
    // @ts-ignore
    useCountdown.mockImplementation(() => '123');

    const { getAllByText } = renderProposalCard({
      votePeriodEnd: '3016-12-12',
    });

    expect(getAllByText('123 left')).toHaveLength(1);
  });

  it('Should render proper timestamp when status Approved', () => {
    // @ts-ignore
    useCountdown.mockImplementation(() => '123');

    const { getAllByText } = renderProposalCard({
      status: 'Approved',
      updatedAt: '3016-12-12',
    });

    expect(getAllByText('12 December 3016')).toHaveLength(1);
  });
});

/* eslint-enable @typescript-eslint/ban-ts-comment */
