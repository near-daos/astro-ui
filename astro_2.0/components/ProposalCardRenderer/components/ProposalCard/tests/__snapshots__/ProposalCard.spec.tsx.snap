// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Proposal Card Should render component 1`] = `
<div>
  <div
    class="root"
    data-testid="proposal-card-root"
  >
    <div
      class="proposalCell"
    >
      <div
        class="root"
      >
        <div
          class="block"
        >
          <div
            class="label"
          >
            proposalCard.proposalTypeVote
          </div>
        </div>
        <div
          class="block"
        >
          <div
            class="valueLarge"
          >
            <div
              class="proposalType"
            >
              viewProposalCard.proposalVariant.poll
              <div
                class="proposalWalletLink"
              >
                <div
                  class="root"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="iconWrapper"
                  >
                    <div>
                      buttonExternal
                    </div>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="countdownCell"
    >
      <span
        class="errorLabel"
      >
        Time expired
      </span>
    </div>
    <div
      class="proposerCell"
    >
      <div
        class="root"
      >
        <div
          class="block"
        >
          <div
            class="label"
          >
            proposalCard.proposalOwner
          </div>
        </div>
        <div
          class="block"
        >
          <div
            class="valueSmall"
          >
            proposer
          </div>
        </div>
      </div>
    </div>
    <div
      class="descriptionCell"
    >
      <div
        class="root fullWidth"
      >
        <div
          class="label"
        >
          <span>
            proposalCard.proposalDescription
            Â 
          </span>
        </div>
        <div
          class="proposalDescription"
        >
          description
        </div>
      </div>
      <div
        class="proposalExternalLink"
      >
        <a
          class="caption1 root"
          href="http://link"
          rel="noreferrer"
          target="_blank"
        >
          <div>
            buttonExternal
          </div>
          <span
            class="body2 text"
          >
            link
          </span>
        </a>
      </div>
    </div>
    <div
      class="contentCell"
    >
      <div>
        Content
      </div>
    </div>
    <div
      class="voteControlCell"
      role="button"
      tabindex="-1"
    >
      <form
        class="root"
      >
        <span
          class="item"
        >
          <button
            class="iconButton large icon"
            disabled=""
            type="submit"
          >
            <div>
              votingYes
            </div>
          </button>
          <span
            class="value title3"
          >
            0
          </span>
        </span>
        <span
          class="item"
        >
          <button
            class="iconButton large icon"
            disabled=""
            type="submit"
          >
            <div>
              votingNo
            </div>
          </button>
          <span
            class="value title3"
          >
            0
          </span>
        </span>
        <span
          class="item"
        >
          <button
            class="iconButton large icon"
            type="button"
          >
            <div>
              chat
            </div>
          </button>
          <span
            class="value title3"
          >
            0
          </span>
        </span>
      </form>
    </div>
    <div
      class="actionBar"
    >
      <div
        class="root"
      >
        <div
          class="root"
        >
          <button
            class="iconButton small rootIcon"
            type="button"
          >
            <div>
              reply
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
